### PR TITLE
Close transaction heartbeats on txn completion.

### LINKED ISSUE #1001: "txn ... abandoned, stopping heartbeat"
```
while true; do
      v=$(dd if=/dev/urandom bs=8 count=1 2>/dev/null | base64)
      ./cockroach kv put $v $v
done
```

gives a lot of messages:

```
level="INFO" time="2015-05-12T13:57:01-04:00" file="txn_coord_sender.go" line="522" msg="transaction \"+a6kSHdDcU4=\":\"\\xd1\\xe9ǝ\\xed\\xc8IO\\x95!\\x0fx\\xd7u\\x04\\xec\" abandoned; stopping heartbeat"
level="INFO" time="2015-05-12T13:57:01-04:00" file="txn_coord_sender.go" line="522" msg="transaction \"IPZaeCmMFjE=\":\"?5\\xc0\\x1ew\\x17O\\xf9\\x9e\\xae\\xd6\\xf2V\\xc7s\\xcb\" abandoned; stopping heartbeat"
level="INFO" time="2015-05-12T13:57:01-04:00" file="txn_coord_sender.go" line="522" msg="transaction \"4wHg6K8X9ZM=\":\"\\xc3\\tj\\xc7b[Eu\\x8c\\x0fHx\\x8e\\a'\\x92\" abandoned; stopping heartbeat"
level="INFO" time="2015-05-12T13:57:01-04:00" file="txn_coord_sender.go" line="522" msg="transaction \"9Qic+bYpSzM=\":\"\\x146ޠ'\\x17B:\\xb5@X\\xac\\xbd\\xea\\x94\\x06\" abandoned; stopping heartbeat"
```

The key-value pairs are there though. It doesn't look like there would be timeouts here, shouldn't the coordinator wrap up when the EndTransaction call comes in?

### PR DESCRIPTION
**Pull request by spencerkimball**
No longer wrap puts and dels in transactions during usage of the kv cli client.

Fixes #1001

### CODE REVIEW

File: kv/txn_coord_sender.go [hunk 1]
```diff
@@ -54,22 +55,27 @@ import (
 // periodically on their own.
 type txnMetadata struct {
 	// txn is the transaction struct from the initial AddRequest call.
-	txn proto.Transaction
```
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: why this change? I'd avoid a pointer here. What would a txnMetadata without a Txn be good for?

File: server/cli/kv.go [hunk 2]
```diff
@@ -122,14 +122,11 @@ func runPut(cmd *cobra.Command, args []string) {
 	if kvDB == nil {
 		return
 	}
-	err := kvDB.Tx(func(tx *client.Tx) error {
```
Comment (bdarnell) [inline server/cli/kv.go]: Why did you make this change? I think that if you operate on multiple keys in one call to the CLI it should be done in one transaction.

File: kv/txn_coord_sender.go [hunk 3]
```diff
@@ -54,22 +55,27 @@ import (
...
+	// atomically.
+	lastUpdateNanos int64
 
 	// timeoutDuration is the time after which the transaction should be
 	// considered abandoned by the client. That is, when
 	// current_timestamp > lastUpdateTS + timeoutDuration If this value
 	// is set to 0, a default timeout will be used.
 	timeoutDuration time.Duration
+
+	// closeCh is closed when the transaction has completed to exit
```
Comment (tamird) [inline kv/txn_coord_sender.go]: something off about this sentence. "closeCh is closed when the transaction completes, which signals that heartbeats should stop"? maybe `closeCh` should be renamed to associated it with heartbeats, too.

**Comment by tschottdorf** [comment]
I actually prefer the client with the txn but mostly because I use it to generate command-line load on the cluster. I'm assuming we'll have other tools for that shortly.

LGTM mod the inline comment

File: kv/txn_coord_sender.go (hunk 1 above)
Comment (spencerkimball) [inline kv/txn_coord_sender.go]: Why make a copy of the transaction? When we init this txnMetadata, the transaction we assign here is already heap-allocated. Might as well just use that.

File: kv/txn_coord_sender.go [hunk 4]
```diff
@@ -296,15 +317,16 @@ func (tc *TxnCoordSender) sendOne(call client.Call) {
 		var txnMeta *txnMetadata
 		if txnMeta, ok = tc.txns[string(header.Txn.ID)]; !ok {
 			txnMeta = &txnMetadata{
-				txn:             *header.Txn,
+				txn:             header.Txn,
```
Comment (bdarnell) [inline kv/txn_coord_sender.go]: This used to make a copy, and now it doesn't. Is there a danger that the caller may reuse the protos after this call and change the data out from under us?

**Comment by bdarnell** [comment]
LGTM

File: server/cli/kv.go (hunk 2 above)
Comment (spencerkimball) [inline server/cli/kv.go]: I was taken aback that tobias' loop would involve transactions at all. I see the argument that this maybe should be all or nothing, but then again, why should you expect different behavior from a command line tool like this between `cockroach kv put a 1 b 2` and `cockroach kv put a 1 && cockroach kv put b 2`? The use of a transaction here wasn't documented consistently and just feels wrong considering we can't otherwise specify them through the command line.

Although it's easy to imagine ways you could find it useful, I really dislike the idea that people would start relying on things like "oh that creates an implicit transaction..." It's complexity where there's no reason to expect it either in the syntax or by convention.

File: kv/txn_coord_sender.go (hunk 4 above)
Comment (spencerkimball) [inline kv/txn_coord_sender.go]: That kind of thing is just problematic in tests, but this is a decent reason to do the copy, so I changed it back.

File: kv/txn_coord_sender.go (hunk 3 above)
Comment (spencerkimball) [inline kv/txn_coord_sender.go]: Renamed and changed comment.

File: kv/txn_coord_sender.go (hunk 4 above)
Comment (bdarnell) [inline kv/txn_coord_sender.go]: I don't think it's just tests - I could see us using a sync.Pool in the rpc server and reusing proto objects that way.
